### DISCUSSION TITLE
coordinates from Polylines in spezified layers to an .txt-File -- solved

### DISCUSSION THREAD

**Discussion by irzp** [mozman/ezdxf#764] [Q&A]
Please, how can I export the coordinates of some dxf-drawings to a table

-from all 2D-Polylines
-sorted by spezified layers
-sorted by closed vs. open

to a simple txt-list
numbered by starting-point to endpoint

for example a rectancgle with the coordinates x/y  p1=(1/1) p2=(1/3), p3=(2/4), p4=(2/2)
1   1   1
2   1   3
3   2   4
4   2   2

I ve tried dxfgrabber but didn't succeed. Will change to ezdxf but have no idea an no skills with python-skripting

greetings gina

**Comment by mozman** [comment]
Everyone has to start somewhere but this is not the forum to get help of this kind. I can't even point you in the direction you should go if you don't have any Python experience.

If you have enough experience with Python, there is documentation on `ezdxf`, some tutorials and many examples to learn from, and if you get stuck then you can ask related questions here.

As on [StackOverflow](https://stackoverflow.com/), I'd like to see your code first to see that you've at least tried to solve the problem. But you can also look for someone on [Fiverr](https://www.fiverr.com/) who will solve your problem for a fee.

**Reply by irzp** [reply]
Thank you for the fast reply,
as asked I send my Python-Code ... 
```Python
# Skript um aus einer DXF-Datei die Koordinaten in eine Text-Datei/ Tabelle zu exportieren
# 2022-11-03

import sys
import ezdxf
import os           # um Dateinamen dieses Python-Skriptes auszulesen
import datetime     # Datum

# Nachdem man die print-Ausgabe am Bildschirm überprüft hat 
# folgende Zeile scharf schalten und   #   entfernen und die Standardausgabe umleiten in
# eine Datei
sys.stdout = open('KOORDINATENLISTE.txt','w')   # a=anhängen, w=überschreiben, t=??!

doc = ezdxf.readfile(sys.argv[1]) #lädt das auszuwertende dxf-File
# doc.saveas("neuangelegt.dxf") #speichert die Originaldatei unter neuem Namen  

now = datetime.datetime.now()
print('KOORDINATENLISTE.txt        Stand: ',str(now))
print('zu dem .dxf-file:', sys.argv[1],"\n")		# Das was hinter dem Skriptname übergeben worden ist

print("(erstellt mit dem Python-Skript:",__file__)     # Dateiname  
print("unter Verwendung der Python-Umgebung   ezdxf ) ","\n\n")     

# ************************************************************

def process_lines(entities):

    for e in entities:
        print("\n")
        print("Layer: %s" % e.dxf.layer)  
        print(str(e))
        # print("Zeichenwerkzeug und #lfd. Nr.: ", str(e))
        dxftype = e.dxftype()
        
        if dxftype == "LINE":          
            print(e.dxf.start)
            print(e.dxf.end)
            # print("Linien-Start (x, y) ", e.dxf.start)
            # print("Linien-Ende (x, y) ", e.dxf.end)

        elif dxftype == "POLYLINE":
            if e.is_2d_polyline or e.is_3D_polyline:
                # other types: is_polygon_mesh and is_poly_face_mesh
                # 2D polylines can have an extrusion != (0, 0, 1) and bulges 
                # see LWPOLYLINE processing.
                for point in e.points():
                    print(point)

        elif dxftype == "LWPOLYLINE":
            # vertices_in_wcs() returns also correct WCS locations for LWPOLYLINE
            # entities placed in 3D space, OCS with extrusion != (0, 0, 1).
            # An extrusion (0, 0, -1) can also happen by mirroring block references!
            # LWPolylines can also contain bulges (arcs), in this case a bulge 
            # is processed as straight line.
            for point in e.vertices_in_wcs():
                print(point)

        elif dxftype == "SPLINE":
            # flatten curve with max distance of 0.01 drawing units between
            # spline and approximation polyline.
            for point in e.flattening(0.01):
                print(point)

def process(entities):
    process_lines(entities)

    # recursive processing of block references:
    for e in entities:
        if e.dxftype() == 'INSERT':
            # create virtual entities based on the block definition
            # and store the content in a list, virtual_entities() returns 
            # a generator. The content will be transformed into the parent 
            # coordinate system:
            process(list(e.virtual_entities()))

# Start processing at the top level: model space
process(doc.modelspace())
```

**Comment by irzp** [comment]
it is OK for my work.
Maybe you can help me forward:
a) sorting the results by layer
b) export of coordinates in an other format ... 
b1) yet it looks like (3253.996934425862, 3094.199219454381, 0.0)
b2) ther shouldn't be brakets around
thank you 
greetings gina

**Comment by mozman** [comment]
Question a) is at least related to `ezdxf` 😄:

For that exist a handy [groupby](https://ezdxf.mozman.at/docs/groupby.html) function/method:

```Python
for layer_name, entities in msp.groupby("layer").items():
    print(f"processing entities of layer '{layer_name}'")
    for e in entities:
        if e.dxftype() == "LINE":
            print(e)
```
The `groupby`  method groups a sequence of DXF entities by a DXF attribute like `layer` and returns a `dict`  with that attribute values as keys and  and a list of entities matching that attribute as values.

The rest of the questions is basic Python knowledge which you have to learn e.g. if you know what a Python `dict` is, you can implement an output sorted by layer that way:

```Python
groups = msp.groupby("layer")
for layer_name in sorted(groups.keys()):
    print(f"processing entities of layer '{layer_name}'")
    for e in groups[layer_name]:
        if e.dxftype() == "LINE":
            print(e)
```

1. Read the docs what data types `ezdxf` returns
2. Learn string formatting in Python
3. Hint: `print(f"{point.x}, {point.y}")` is a starting point ...

**Reply by irzp** [reply]
thanks a lot
